halo2_frontend/src/plonk/circuit/constraint_system.rs

```diff
@@ -404,6 +410,14 @@ impl<F: Field> ConstraintSystem<F> {
                 if table.contains_simple_selector() {
                     panic!("expression containing simple selector supplied to lookup argument");
                 }
+                // this check ensures that one extra, dedicated `Fixed` column or `Selector` is used
+                // for enabling the real table rows of the column, which is used as `TableColumn`.
+                // otherwise, we get the soundness error, like https://github.com/privacy-scaling-explorations/halo2/issues/335
+                if table.degree() < 2 {
```

Comment:
I think this check is not effective in certain scenarios.
- For example, let's say you build a table expression by multiplying two advice queries.  The degree is 2, but it's still problematic.  This would be a false negative.
- Imagine you have a dynamic table that combines fixed and advice columns.  In that case it can be OK for some table expressions to be degree 1.  This would be a false positive.

As an example of the second case, take a look at the `block_table` from the zkEVM CE project: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#block_table
The `tag` and `index` columns are fixed and the `valueLo` and `valueHi` are advice.
Assume that the `tag` values go from 1 to N (and the assignments to the unused rows are 0)

now we want to do a lookup to `Time` for example.  The input expression would be `[q * Time, 0, q * in_value, 0]` and the table expression would be `[Time, 0, value, 0]`.  This table expression can only match a valid row, because invalid rows have Tag=0 so they are `[0, ... ]`, and `Time != 0`.

So I think the proper check would be to make sure that at least 1 fixed query exists in some of the table expressions.  Then we must forbid that all table expressions have 0 fixed queries.

Reply:
Ah, thanks for pointing out specific cases.
Yes, I totally agree with your idea.
Will improve the check.